File: Assignment1/version1/Java/Application.java
package individual.v1;

import java.io.*;
import java.util.Scanner;

//Class for applicatoin
public class Application {
	Scanner kb = new Scanner(System.in);
	Set[] neighbors;

	// introduce Bron-Kerbosh algorithm
	// create grpah from user input
	// represent graph as array of sets
	public void intro() {
		System.out.println("Welcome to the Bron-Kerbosh algorithm.\n"
							+ "This algorithm finds the maximal cliques within an undirected graph using sets.\n"
							+ "Enter number of nodes within graph");
		
		int nodes;
		while(true){
			try{
				nodes = kb.nextInt();
				break;
			}catch(Exception e){
				System.out.println("Enter a number");
			}	
		}

		neighbors = new Set[nodes];

		System.out.print("Graph nodes: ");
		for (int i = 0; i < nodes; i++) {
			neighbors[i] = new Set();
			System.out.print(i + " ");
		}
		System.out.println("\nEnter the edges of the graphs.\n"
						+ "to set an edge, enter the two nodes individually.");
		String cont = "y";
		
		//Get input to create edge
		while (cont.equalsIgnoreCase("y")){
			System.out.println("first node index: ");
			int first = kb.nextInt();
			System.out.println("Second node index: ");
			int second = kb.nextInt();

			neighbors[first].addElement(Integer.toString(second));
			neighbors[second].addElement(Integer.toString(first));

			System.out.println("Edge between node " + first
					+ " and " + second + " has been made");
			kb.nextLine();
			System.out.println("Continue? y/n");
			cont = kb.nextLine();
			//Ask user if they wish to continue adding edge
			//Loop question if answer is not y/n
			while(!cont.equalsIgnoreCase("y") && !cont.equalsIgnoreCase("n")){
				System.out.println("Invalid input enter y/n");
				cont = kb.nextLine();
			}
		}
	}

	// method that implements the actual Bron-Kerb
	public void bronKerb(Set a, Set b, Set c) {
		if (b.size == 0 && c.size == 0) {
			System.out.print("Maximal Clique: ");
			a.printSet();
		} 
		//This else if is supposed to be empty
		else if (b.size == 0 && c.size != 0) {
		} 
		
		else {
			int count = b.set.length;
			while (b.size != 0) {
				Set v = new Set(b.set[0]);

				Set newR = a.union(v);
				Set newP = b.intersection(neighbors[Integer.parseInt(b.set[0])]);
				Set newX = c.intersection(neighbors[Integer.parseInt(b.set[0])]);

				bronKerb(newR, newP, newX);
				b = b.difference(v);
				c = c.union(v);
			}
		}

	}

}

File: CS408 Project1/version1/C++/Application.cpp
#include <string>
#include <vector>
#include "Set.h"
#include "Application.h"
#include <iostream>
#include <algorithm>

using namespace std;

//.cpp file for application class
Application::Application() {
}

//Intro method that explain what algorithm will be performed
//Also set up the necessary vector and edges
void Application::intro() {
	cout << "Welcome to the Bron-Kerbosh algorithm.\n"
		"This algorithm finds the maximal cliques within an undirected graph using sets.\n";

	int nodes;

	//Get number of nodes
	while (true) {
		cout << "Enter number of nodes within graph: ";
		cin >> nodes;

		//Catch input mismatch
		if (cin.fail() || nodes < 0 || nodes < 2) {
			cin.clear();
			cin.ignore(numeric_limits <streamsize>::max(), '\n');
			cout << "Invalid input" << endl;
		}
		else {
			break;
		}
	}

	cout << "Graph nodes: ";

	for (int i = 0; i < nodes; i++) {
		neighbors.push_back(Set());
		cout << i ;
		cout << " ";
	}
	
	cout << "\nEnter the edges of the graphs.\n"
			 "to set an edge, enter the two nodes individually." << endl;
	
	string cont = "y";

	int first = 0;
	int second = 0;

	//Get input to create edge
	while (!cont.compare("y")) {
		while (true) {
			cout << "first node index: ";
			cin >> first;
			//Catch input mismatch
			if (cin.fail()) {
				cin.clear();
				cin.ignore(numeric_limits <streamsize>::max(), '\n');
				cout << "Invalid input" << endl;
				continue;
			}
			else if (first >= nodes || first < 0) {
				cout << "Input out of index" << endl;
				continue;
			}
			
			cout << "Second node index: ";
			cin >> second;
			//Catch input mismatch
			if (cin.fail()) {
				cin.clear();
				cin.ignore(numeric_limits <streamsize>::max(), '\n');
				cout << "Invalid input" << endl;
				continue;
			}
			else if (second >= nodes || second < 0) {
				cout << "Input out of index" << endl;
				continue;
			}
			
			if (first == second){
				cout << "Must enter different nodes" << endl;
				continue;
			}			
			
			break;
		}

		//add edges
		neighbors.at(first).addElement(to_string(second));
		neighbors.at(second).addElement(to_string(first));

		printf("Edge between node %d and %d has been made\n", first, second);
		
		cout << "Continue? y/n: ";
		cin >> cont;
		transform(cont.begin(), cont.end(), cont.begin(), ::tolower);
		//Ask user if they wish to continue adding edge
		//Loop question if answer is not y/n
		while (cont.compare("y") && cont.compare("n")) {
			cout << "Invalid input enter y/n: ";
			cin >> cont;
			transform(cont.begin(), cont.end(), cont.begin(), ::tolower);
		}
	}
}

//Method that imlements the actual Bron-Kerb
void Application::bronKerb(Set a, Set b, Set c) {
	//Maximal clique found
	if (b.getSize() == 0 && c.getSize() == 0) {
		cout << "Maximal clique: ";
		a.printSet();
	}

	//This else if is supposed to be empty
	else if (b.getSize() == 0 && c.getSize() != 0) {
	}

	else {
		int count = b.getSize();
		while (b.getSize() != 0) {
			Set v(b.getSet().at(0));

			Set newA = a + v;
			Set newB = b ^ neighbors.at(stoi(b.getSet().at(0)));
			Set newC = c ^ neighbors.at(stoi(b.getSet().at(0)));

			bronKerb(newA, newB, newC);
			b = b - v;
			c = c + v;
		}
	}
}

int Application::getSize() {
	return neighbors.size();
}